### PR TITLE
[Flow decomposition][Reformat] Observer and NetworkUtil

### PR DESCRIPTION
**Pull request by OpenSuze**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

The goal is to remove terminal setP and setQ in losses compensation PR #147.
I used this PR to reformat observers in flow decompostion computer and the NetworkUtils functions.

**What kind of change does this PR introduce?**

Reformating

**What is the current behavior?**

No change

**What is the new behavior (if this is a feature change)?**
No change

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [x] No

**Other information**:

### CODE REVIEW

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java [hunk 1]
```diff
@@ -46,25 +59,10 @@ public static Country getTerminalCountry(Terminal terminal) {
...
-                Function.identity()
-            ));
-    }
-
-    public static Country getInjectionCountry(Injection<?> injection) {
-        return getTerminalCountry(injection.getTerminal());
+            .collect(Collectors.toMap(idList::get, Function.identity()));
     }
 
     public static List<Branch> getAllValidBranches(Network network) {
```
Comment (caioluke) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:68]: Is the TODO from `.filter(NetworkUtil::isInMainSynchronousComponent)` still valid? It is from 2022

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java [hunk 2]
```diff
@@ -83,11 +81,63 @@ private static boolean isInMainSynchronousComponent(Branch<?> branch) {
...
+        return !(injection instanceof DanglingLine danglingLine && danglingLine.isPaired());
+    }
+
+    private static boolean isInjectionInMainSynchronousComponent(Injection<?> injection) {
+        return NetworkUtil.isTerminalInMainSynchronousComponent(injection.getTerminal());
+    }
+
+    public static List<String> getPstIdList(Network network) {
+        return network.getTwoWindingsTransformerStream()
+            .filter(NetworkUtil::isPst)
```
Comment (caioluke) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:130]:
```suggestion
            .filter(PhaseTapChangerHolder::hasPhaseTapChanger)
```

And remove the `isPst` function

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java [hunk 3]
```diff
@@ -83,11 +81,63 @@ private static boolean isInMainSynchronousComponent(Branch<?> branch) {
...
+    private static boolean isNotPairedDanglingLine(Injection<?> injection) {
+        return !(injection instanceof DanglingLine danglingLine && danglingLine.isPaired());
+    }
+
+    private static boolean isInjectionInMainSynchronousComponent(Injection<?> injection) {
+        return NetworkUtil.isTerminalInMainSynchronousComponent(injection.getTerminal());
+    }
+
+    public static List<String> getPstIdList(Network network) {
+        return network.getTwoWindingsTransformerStream()
```
Comment (caioluke) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:129]: What about ThreeWindingsTransformers? Shouldn't we add them too?
If not, shouldn't we warn the user about that?

**Review by caioluke — COMMENTED** [review]
Thank you for the refactoring and the utils tests. 😄

I think observers are much better placed now!

Question remains about the 2 TODO's and whether or not to include three windings transfomers in `getPstIdList`.

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java [hunk 4]
```diff
@@ -83,11 +81,63 @@ private static boolean isInMainSynchronousComponent(Branch<?> branch) {
...
+        return network.getDanglingLineStream()
+            .filter(dl -> !dl.isPaired())
+            .filter(NetworkUtil::isInjectionConnected)
+            .filter(NetworkUtil::isInjectionInMainSynchronousComponent)
+            .map(danglingLine -> (Injection<?>) danglingLine)
+            .collect(Collectors.toList());
+    }
+
+    private static boolean managedInjectionTypes(Injection<?> injection) {
+        return !(injection instanceof BusbarSection || injection instanceof ShuntCompensator || injection instanceof StaticVarCompensator); // TODO Remove this fix once the active power computation after a DC load flow is fixed in OLF
```
[the remark this answers is not in the document]
Comment (OpenSuze) [reply flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java]: Yes, I added some tests to highlight this. I will also write documentation

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java (hunk 1 above)
Comment (OpenSuze) [reply flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:69]: Yes, I added some tests to highlight this. I will also write documentation

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java (hunk 3 above)
Comment (OpenSuze) [reply flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:150]: Do you have a methodology to decompose their flow ? ThreeWindingsTransformers have a PhaseTapChanger per leg...

Comment (OpenSuze) [reply flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:150]: But, we could write some documentation about them

**Comment by OpenSuze** [comment]
I added a lot of test with some "golden master" tests

**Comment by OpenSuze** [comment]
- [ ] Check new TODO
- [ ] Write documentation injection != BusbarSection ShuntCompensator StaticVarCompensator
- [ ] Write documentation main synchronous component
- [ ] Write documentation ThreeWindingsTransformers
- [ ] merge with losses_compensation_improvement branch

**Comment by OpenSuze** [comment]
Thank you for the review, I made some changes. Ready for review

**Review by caioluke — APPROVED** [review]
Wow, thanks a lot for the work you put into this PR :)

I have only some minor comments

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java [hunk 5]
```diff
@@ -40,54 +54,114 @@ public static Country getTerminalCountry(Terminal terminal) {
...
+    }
+
+    private static boolean isNotAStaticVarCompensatorAsTheyAreNotValidInjectionsForSensitivityComputation(Injection<?> injection) {
+        return !(injection instanceof StaticVarCompensator);
+    }
+
+    public static List<String> getPstIdList(Network network) {
+        return network.getTwoWindingsTransformerStream()
+            .filter(NetworkUtil::isConnected)
+            .filter(NetworkUtil::hasPhaseTapChanger)
```
[the remark this answers is not in the document]
Comment (OpenSuze) [reply flow-decomposition/src/main/java/com/powsybl/flow_decomposition/NetworkUtil.java:152]: Thank you ! I did not know that :)

File: flow-decomposition/src/test/java/com/powsybl/flow_decomposition/TestUtils.java [hunk 6]
```diff
@@ -43,12 +48,22 @@ public static void assertCoherenceTotalFlow(FlowDecompositionParameters.RescaleM
...
             }
         }
     }
 
-    static double getLossOnBus(Network network, Bus bus) {
+    private static void assertEqualsWithoutRescaling(String xnec, DecomposedFlow decomposedFlow) {
+        if (Double.isNaN(decomposedFlow.getDcReferenceFlow())) {
+            LOGGER.error("XNEC \"{}\" is probably not connected", xnec); // TODO: should we decompose such xnecs ?
+        } else if (decomposedFlow.getTotalFlow() == 0) {
+            LOGGER.error("XNEC \"{}\" is outside main synchronous component", xnec); // TODO: should we decompose such xnecs ?
```
[the remark this answers is not in the document]
Comment (OpenSuze) [reply flow-decomposition/src/test/java/com/powsybl/flow_decomposition/TestUtils.java:60]: Created https://github.com/powsybl/powsybl-entsoe/issues/150

**Comment by OpenSuze** [comment]
Waiting for a final green light to merge this PR

**Comment by caioluke** [comment]
@OpenSuze it's all good for me!

**Pull request merged**